src/Plugin/Commerce/PaymentGateway/ExpressCheckout.php

```diff
@@ -0,0 +1,570 @@
+        $extra['refund_type'] = 'Full';
+      }
+    }
+
+    // RefundTransaction API Operation (NVP).
+    // Refund (full or partial) an Express Checkout transaction.
+    $paypal_response = $this->apiRefundTransaction($payment, $extra);
+
+    if ($paypal_response['ACK'] == 'Failure') {
+      $message = $paypal_response['L_LONGMESSAGE0'];
+      throw new PaymentGatewayException($message, $paypal_response['L_ERRORCODE0']);
+    }
+
+    $payment->setRefundedAmount($new_refunded_amount);
+    $payment->save();
+  }
+
+  /**
+   * PayPal Express Checkout NVP API helper methods.
+   */
+
+  /**
+   * {@inheritdoc}
+   */
+  public function apiSetExpressCheckout(PaymentInterface $payment, $extra) {
+    $order = $payment->getOrder();
+
+    $amount = $payment->getAmount()->getNumber();
+    $configuration = $this->getConfiguration();
+
+    if ($extra['capture']) {
+      $payment_action = 'Sale';
+    }
+    else {
+      $payment_action = 'Authorization';
+    }
+
+    $flow = 'ec';
+    // Build a name-value pair array for this transaction.
+    $nvp_data = [
+      'METHOD' => 'SetExpressCheckout',
+
+      // Default the Express Checkout landing page to the Mark solution.
+      'SOLUTIONTYPE' => 'Mark',
+      'LANDINGPAGE' => 'Login',
+
+      // Disable entering notes in PayPal, as we don't have any way to accommodate
+      // them right now.
+      'ALLOWNOTE' => '0',
+
+      'PAYMENTREQUEST_0_PAYMENTACTION' => $payment_action,
+      'PAYMENTREQUEST_0_AMT' => $this->formatNumber($amount),
+      'PAYMENTREQUEST_0_CURRENCYCODE' => $payment->getAmount()->getCurrencyCode(),
+      'PAYMENTREQUEST_0_INVNUM' => $order->id() . '-' . REQUEST_TIME,
+
+      // Set the return and cancel URLs.
+      'RETURNURL' => $extra['return_url'],
+      'CANCELURL' => $extra['cancel_url'],
+    ];
+
+    $order_express_checkout_data = $order->getData('paypal_express_checkout');
+    if (!empty($order_express_checkout_data['token'])) {
+      $nvp_data['TOKEN'] = $order_express_checkout_data['token'];
+    }
+
+    // Get the order line items
+    $n = 0;
+    foreach ($order->getItems() as $item) {
+      $item_amount = $item->getUnitPrice()->getNumber();
+      $nvp_data += [
+        'L_PAYMENTREQUEST_0_NAME' . $n => $item->getTitle(),
+        'L_PAYMENTREQUEST_0_AMT' . $n => $this->formatNumber($item_amount),
+        'L_PAYMENTREQUEST_0_QTY' . $n => $item->getQuantity(),
+      ];
+      $n++;
+    }
+
+    // If reference transactions are enabled and a billing agreement is supplied...
+    if (!empty($configuration['reference_transactions']) &&
+      !empty($configuration['ba_desc'])) {
+      $nvp_data['BILLINGTYPE'] = 'MerchantInitiatedBillingSingleAgreement';
+      $nvp_data['L_BILLINGTYPE0'] = 'MerchantInitiatedBillingSingleAgreement';
+      $nvp_data['L_BILLINGAGREEMENTDESCRIPTION0'] = $configuration['ba_desc'];
+    }
+
+    // If Express Checkout Account Optional is enabled...
+    if ($configuration['solution_type'] != 'Mark') {
```

Comment:
~~Isn't the solution type a radio? So we could just default to this value above?~~

I see. So Sole/SoleBilling splits into solution type and dictates landing page.